examples/dqn/dqn_atari.py

```diff
@@ -0,0 +1,338 @@
+"""
+DQN Benchmarks: Reproducing Experiments from Mnih et al. 2015
+Deep Q-Learning Algorithm on Atari Environments.
+"""
+
+import random
+import time
+
+import gym
+import hydra
+import numpy as np
+import torch.nn
+import torch.optim
+import tqdm
+from tensordict import TensorDict
+
+from torchrl.collectors import SyncDataCollector
+from torchrl.data import CompositeSpec, LazyMemmapStorage, TensorDictReplayBuffer
+from torchrl.envs import (
+    CatFrames,
+    default_info_dict_reader,
+    DoubleToFloat,
+    EnvCreator,
+    ExplorationType,
+    GrayScale,
+    NoopResetEnv,
+    ParallelEnv,
+    Resize,
+    RewardClipping,
+    RewardSum,
+    SerialEnv,
+    set_exploration_type,
+    StepCounter,
+    ToTensorImage,
+    TransformedEnv,
+    VecNorm,
+)
+from torchrl.envs.libs.gym import GymWrapper
+from torchrl.modules import ConvNet, EGreedyWrapper, MLP, QValueActor
+from torchrl.objectives import DQNLoss, HardUpdate
+from torchrl.record.loggers import generate_exp_name, get_logger
+
+# ====================================================================
+# Environment utils
+# --------------------------------------------------------------------
+
+
+class EpisodicLifeEnv(gym.Wrapper):
+    def __init__(self, env):
+        """Make end-of-life == end-of-episode, but only reset on true game over.
+        Done by DeepMind for the DQN and co. It helps value estimation.
+        """
+        gym.Wrapper.__init__(self, env)
+        self.lives = 0
+
+    def step(self, action):
+        obs, rew, done, info = self.env.step(action)
+        lives = self.env.unwrapped.ale.lives()
+        info["end_of_life"] = False
+        if (lives < self.lives) or done:
+            info["end_of_life"] = True
+        self.lives = lives
+        return obs, rew, done, info
+
+    def reset(self, **kwargs):
+        reset_data = self.env.reset(**kwargs)
+        self.lives = self.env.unwrapped.ale.lives()
+        return reset_data
+
+
+# TODO: this function makes collection crash
+# def make_env(env_name, device, is_test=False):
+#     env = gym.make(env_name)
+#     if not is_test:
+#         env = EpisodicLifeEnv(env)
+#     env = GymWrapper(
+#         env, frame_skip=frame_skip, from_pixels=True, pixels_only=False, device=device
+#     )
+#     env = TransformedEnv(env)
+#     reader = default_info_dict_reader(["end_of_life"])
+#     env.set_info_dict_reader(reader)
+#     env.append_transform(NoopResetEnv(noops=8))
+#     env.append_transform(ToTensorImage())
+#     env.append_transform(GrayScale())
+#     env.append_transform(Resize(84, 84))
+#     env.append_transform(CatFrames(N=4, dim=-3))
+#     env.append_transform(RewardSum())
+#     env.append_transform(StepCounter(max_steps=4500))
+#     if not is_test:
+#         env.append_transform(RewardClipping(-1, 1))
+#     env.append_transform(DoubleToFloat())
+#     # env.append_transform(VecNorm(in_keys=["pixels"]))
+#     return env
```

Comment:
What makes it crash? Anything we can fix?

Reply:
This it what makes it crash: https://github.com/pytorch/rl/issues/1518